transport/http/jsonrpc/server.go

```diff
@@ -0,0 +1,202 @@
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	httptransport "github.com/go-kit/kit/transport/http"
+)
+
+// Server wraps an endpoint and implements http.Handler.
+type Server struct {
+	ctx       context.Context
+	ecm       EndpointCodecMap
+	before    []httptransport.RequestFunc
+	after     []httptransport.ServerResponseFunc
+	finalizer httptransport.ServerFinalizerFunc
+	logger    log.Logger
+}
+
+// NewServer constructs a new server, which implements http.Server.
+func NewServer(
+	ctx context.Context,
+	ecm EndpointCodecMap,
+	options ...ServerOption,
+) *Server {
+	s := &Server{
+		ctx:    ctx,
+		ecm:    ecm,
+		logger: log.NewNopLogger(),
+	}
+	for _, option := range options {
+		option(s)
+	}
+	return s
+}
+
+// EndpointCodec defines and Endpoint and its associated codecs
+type EndpointCodec struct {
+	Endpoint endpoint.Endpoint
+	Decode   DecodeRequestFunc
+	Encode   EncodeResponseFunc
+}
+
+// EndpointCodecMap maps the Request.Method to the proper EndpointCodec
+type EndpointCodecMap map[string]EndpointCodec
+
+// ServerOption sets an optional parameter for servers.
+type ServerOption func(*Server)
+
+// ServerBefore functions are executed on the HTTP request object before the
+// request is decoded.
+func ServerBefore(before ...httptransport.RequestFunc) ServerOption {
+	return func(s *Server) { s.before = before }
+}
+
+// ServerAfter functions are executed on the HTTP response writer after the
+// endpoint is invoked, but before anything is written to the client.
+func ServerAfter(after ...httptransport.ServerResponseFunc) ServerOption {
+	return func(s *Server) { s.after = after }
+}
+
+// ServerErrorLogger is used to log non-terminal errors. By default, no errors
+// are logged. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ServerErrorEncoder or ServerFinalizer, both of which have access to
+// the context.
+func ServerErrorLogger(logger log.Logger) ServerOption {
+	return func(s *Server) { s.logger = logger }
+}
+
+// ServerFinalizer is executed at the end of every HTTP request.
+// By default, no finalizer is registered.
+func ServerFinalizer(f httptransport.ServerFinalizerFunc) ServerOption {
+	return func(s *Server) { s.finalizer = f }
+}
+
+// ServeHTTP implements http.Handler.
+func (s Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, "405 must POST\n")
+		return
+	}
+	ctx := s.ctx
+
+	if s.finalizer != nil {
+		iw := &interceptingWriter{w, http.StatusOK}
+		defer func() { s.finalizer(ctx, iw.code, r) }()
+		w = iw
+	}
+
+	for _, f := range s.before {
+		ctx = f(ctx, r)
+	}
+
+	// Decode the body into an  object
+	var req Request
+	err := json.NewDecoder(r.Body).Decode(&req)
```

Comment:
It's a bit troubling that there is no way to wrap the r.Body by another io.Reader here. For example someone might want to wrap it with a `io.LimitReader` to ensure that the server is not flooded with a stream of neverending bytes. 

Perhaps a ServerOption can be added here.